web/src/engine/osk/src/visualKeyboard.ts

```diff
@@ -488,42 +502,41 @@ export default class VisualKeyboard extends EventEmitter<EventMap> implements Ke
         // First, if we've configured the gesture to generate a keystroke, let's handle that.
         const gestureKey = gestureStage.item;
 
-        let coordSource = gestureStage.sources[0];
-        let coord: InputSample<KeyElement, string> = null;
-        if(coordSource) {
-          // TODO:  should probably vary depending upon `gestureStage.matchedId`
-          // (certain types should probably use the base coord... or even from
-          // a prior stage of the sequence as appropriate.)
-          //
-          // This is the coordinate used as the basis for fat-finger calculations.
-          coord = coordSource.currentSample;
-        }
+        const coordSource = gestureStage.sources[0];
+        const coord: InputSample<KeyElement, string> = coordSource ? coordSource.currentSample : null;
 
-        let keyResult: {
-          contextToken?: number
-        } = null;
+        let keyResult: KeyRuleEffects = null;
 
         // Multitaps do special key-mapping stuff internally and produce + raise their
         // key events directly.
-        if(gestureKey && !(handler instanceof Multitap)) {
+        if(gestureKey && !(handlers && handlers[0] instanceof Multitap)) {
           let correctionKeyDistribution: KeyDistribution;
-          const baseDistanceMap = this.getSimpleTapCorrectionDistances(coord, gestureKey.key.spec as ActiveKey);
+          const baseDistanceMap = this.getSimpleTapCorrectionDistances(coordSource.currentSample, gestureKey.key.spec as ActiveKey);
 
-          if(handler) {
+          if(handlers) {
             // Certain gestures (especially flicks) like to consider the base layout as part
             // of their corrective-distribution calculations.
             //
             // May be `null` for gestures that don't need custom correction handling,
             // such as modipresses or initial/simple-tap keystrokes.
-            correctionKeyDistribution = handler.currentStageKeyDistribution(baseDistanceMap);
+            correctionKeyDistribution = handlers[0].currentStageKeyDistribution(baseDistanceMap);
           }
 
           if(!correctionKeyDistribution) {
             correctionKeyDistribution = distributionFromDistanceMaps(baseDistanceMap);
           }
 
           // Once the best coord to use for fat-finger calculations has been determined:
-          keyResult = this.modelKeyClick(gestureStage.item, coord);
+          const shouldLockLayer = handlers && (handlers[0] instanceof SubkeyPopup) && handlers[0].shouldLockLayer;
+          try {
+            if(shouldLockLayer) {
+              this.lockLayer(true);
+            }
+            keyResult = this.modelKeyClick(gestureStage.item, coord);
+          } finally {
+            this.lockLayer(false);
```

Comment:
Should this be matching the earlier statement? For consistency?
```suggestion
            if(shouldLockLayer) {
              this.lockLayer(false);
            }
```